language/tools/move-bytecode-utils/src/layout.rs

```diff
@@ -158,55 +174,115 @@ impl<'a, T: GetModule> SerdeLayoutBuilder<'a, T> {
         module_id: &ModuleId,
         name: &Identifier,
         type_arguments: &[Format],
-        shallow: bool,
     ) -> Result<Format, T::Error> {
         // build a human-readable name for the struct type. this should do the same thing as
         // StructTag::display(), but it's not easy to use that code here
-        let struct_key = if type_arguments.is_empty() {
-            format!("{}::{}", module_id, name)
+
+        let declaring_module = self
+            .module_resolver
+            .get_module_by_id(module_id)?
+            .expect("Failed to resolve module");
+        let def = declaring_module
+            .borrow()
+            .find_struct_def_by_name(name)
+            .unwrap_or_else(|| {
+                panic!(
+                    "Could not find struct named {} in module {}",
+                    name,
+                    declaring_module.borrow().name()
+                )
+            });
+        let normalized_struct = Struct::new(declaring_module.borrow(), def).1;
+        assert_eq!(
+            normalized_struct.type_parameters.len(),
+            type_arguments.len(),
+            "Wrong number of type arguments for struct"
+        );
+
+        let generics: Vec<String> = type_arguments
+            .iter()
+            .zip(normalized_struct.type_parameters.iter())
+            .filter_map(|(type_arg, type_param)| {
```

Comment:
minor nit, I like `filter_map`, but I feel like this would be more readable with distinct `filter` and `map` calls, since neither the filter function nor the mapping function inherently return an `Option`

Reply:
I found the conditional logic harder to write/understand by using the filter/map, so left it as is.